imap_processing/hi/l1c/hi_l1c.py

```diff
@@ -0,0 +1,221 @@
+    cdf_manager.load_variable_attributes("imap_hi_variable_attrs.yaml")
+
+    # preallocate coordinates xr.DataArrays
+    coords = dict()
+    # epoch coordinate has only 1 entry for pointing set
+    attrs = cdf_manager.get_variable_attributes("hi_pset_epoch").copy()
+    dtype = attrs.pop("dtype")
+    coords["epoch"] = xr.DataArray(
+        np.empty(1, dtype=dtype),
+        name="epoch",
+        dims=["epoch"],
+        attrs=attrs,
+    )
+    attrs = cdf_manager.get_variable_attributes("hi_pset_esa_step").copy()
+    dtype = attrs.pop("dtype")
+    coords["esa_step"] = xr.DataArray(
+        np.full(n_esa_steps, attrs["FILLVAL"], dtype=dtype),
+        name="esa_step",
+        dims=["esa_step"],
+        attrs=attrs,
+    )
+    # spin angle bins are 0.1 degree bins for full 360 degree spin
+    attrs = cdf_manager.get_variable_attributes("hi_pset_spin_angle_bin").copy()
+    dtype = attrs.pop("dtype")
+    coords["spin_angle_bin"] = xr.DataArray(
+        np.arange(int(360 / 0.1), dtype=dtype),
+        name="spin_angle_bin",
+        dims=["spin_angle_bin"],
+        attrs=attrs,
+    )
+
+    # Allocate the variables
+    data_vars = dict()
+    # despun_z is a 1x3 unit vector that does not have a DEPEND_1.
+    # Define this dict to override the shape produced in full_dataarray
+    var_shapes = {"despun_z": (1, 3)}
+    for var_name in [
+        "despun_z",
+        "hae_latitude",
+        "hae_longitude",
+        "counts",
+        "exposure_times",
+        "background_rates",
+        "background_rates_uncertainty",
+    ]:
+        data_vars[var_name] = full_dataarray(
+            var_name,
+            cdf_manager.get_variable_attributes(f"hi_pset_{var_name}"),
+            coords,
+            shape=var_shapes.get(var_name, None),
+        )
+
+    # Generate label variables
+    data_vars["esa_step_label"] = xr.DataArray(
+        coords["esa_step"].values.astype(str),
+        name="esa_step_label",
+        dims=["esa_step"],
+        attrs=cdf_manager.get_variable_attributes("hi_pset_esa_step_label"),
+    )
+    data_vars["spin_bin_label"] = xr.DataArray(
+        coords["spin_angle_bin"].values.astype(str),
+        name="spin_bin_label",
+        dims=["spin_angle_bin"],
+        attrs=cdf_manager.get_variable_attributes("hi_pset_spin_bin_label"),
+    )
+    data_vars["label_vector_HAE"] = xr.DataArray(
+        np.array(["x HAE", "y HAE", "z HAE"], dtype=str),
+        name="label_vector_HAE",
+        dims=[" "],
+        attrs=cdf_manager.get_variable_attributes("hi_pset_label_vector_HAE"),
+    )
+
+    pset_global_attrs = cdf_manager.get_global_attributes(
+        "imap_hi_l1c_pset_attrs"
+    ).copy()
+    pset_global_attrs["Logical_source"] = pset_global_attrs["Logical_source"].format(
+        sensor=sensor_str
+    )
+    dataset = xr.Dataset(data_vars=data_vars, coords=coords, attrs=pset_global_attrs)
+    return dataset
+
+
+def full_dataarray(name, attrs, coords: dict, shape=None):
```

Comment:
This is a really nice helper! For later: I think we should look at moving this out more generally and using it across instruments if possible.